### ISSUE TITLE
[clang-tidy] cppcoreguidelines-pro-bounds-avoid-unchecked-container-access: false positive when index is bounds-checked in same scope

### ISSUE THREAD

**Issue by raidenluikang** [llvm/llvm-project#213664]
See following case: 

```c++
#include <iostream>
#include <vector>

int main(int argc, char*  /*argv*/[]) {
    std::vector<int> vec;
    vec.push_back(1);
    vec.push_back(2);
    vec.push_back(3);
    if (argc < 3)
        argc = 4;

    if (argc >= 0 && static_cast<size_t>(argc) < vec.size()) {
        std::cout << vec[argc] << std::endl; // NOLINT(*-pro-bounds-avoid-unchecked-container-access)
    } else {
        std::cout << "Index of bound " << std::endl;
    }
    return 0;
}
```

call `vec[argc]` - should be safe, but we must ignore it with NOLINT.

```console
$clang-tidy --version:
  LLVM (http://llvm.org/):
  LLVM version 23.0.0git
  Optimized build.
```

IDE:  CLion 2026.1.2

https://clang.llvm.org/extra/clang-tidy/checks/bugprone/unchecked-optional-access.html  
For optional this is safe case:

```c++
void f(std::optional<int> opt) {
  if (opt.has_value()) {
    use(*opt);
  }
}
```

so, why `vector::operator[] (size_t index)` can't check same as?

**Comment by vbvictor** [comment]
Those check are different by design.
cppcoreguidelines-pro-bounds-avoid-unchecked-container-access simply flags all `[]` regarding of its safeness.

bugprone-unchecked-optional-access tries to find only provable "bad" accesses.

see docs from cppcoreguidelines-pro-bounds-avoid-unchecked-container-access:

```
Finds calls to operator[] in STL containers and suggests replacing them with safe alternatives. Safe alternatives include STL at or GSL at functions, begin() or end() functions, range-for loops, std::span, or an appropriate function from <algorithms>.
```

If it doesn't suit you - I suggest disabling the check
